internal/graph/lr2streams.go

```diff
@@ -0,0 +1,290 @@
+		lock:               &sync.Mutex{},
+	}
+
+	return rdc.runAndWait()
+}
+
+type rdc struct {
+	parentRequest      ValidatedLookupResources2Request
+	foundResources     dispatchableResourcesSubjectMap2
+	ci                 cursorInformation
+	parentStream       dispatch.LookupResources2Stream
+	newSubjectType     *core.RelationReference
+	filteredSubjectIDs []string
+	entrypoint         typesystem.ReachabilityEntrypoint
+	lrDispatcher       dispatch.LookupResources2
+	checkDispatcher    dispatch.Check
+
+	taskrunner *taskrunner.TaskRunner
+
+	lock *sync.Mutex
+}
+
+func (rdc *rdc) dispatchAndCollect(ctx context.Context, cursor *v1.Cursor) ([]*v1.DispatchLookupResources2Response, error) {
+	collectingStream := dispatch.NewCollectingDispatchStream[*v1.DispatchLookupResources2Response](ctx)
+	err := rdc.lrDispatcher.DispatchLookupResources2(&v1.DispatchLookupResources2Request{
+		ResourceRelation: rdc.parentRequest.ResourceRelation,
+		SubjectRelation:  rdc.newSubjectType,
+		SubjectIds:       rdc.filteredSubjectIDs,
+		TerminalSubject:  rdc.parentRequest.TerminalSubject,
+		Metadata: &v1.ResolverMeta{
+			AtRevision:     rdc.parentRequest.Revision.String(),
+			DepthRemaining: rdc.parentRequest.Metadata.DepthRemaining - 1,
+		},
+		OptionalCursor: cursor,
+		OptionalLimit:  uint32(datastore.FilterMaximumIDCount),
+	}, collectingStream)
+	return collectingStream.Results(), err
+}
+
+func (rdc *rdc) runDispatch(ctx context.Context, cursor *v1.Cursor) error {
+	rdc.lock.Lock()
+	if rdc.ci.limits.hasExhaustedLimit() {
+		rdc.lock.Unlock()
+		return nil
+	}
+	rdc.lock.Unlock()
+
+	collected, err := rdc.dispatchAndCollect(ctx, cursor)
+	if err != nil {
+		return err
+	}
+
+	if len(collected) == 0 {
+		return nil
+	}
+
+	// Kick off a worker to filter the results via a check and then publish what was found.
+	rdc.taskrunner.Schedule(func(ctx context.Context) error {
+		return rdc.runChecker(ctx, collected)
+	})
+
+	// Start another dispatch at the cursor of the last response, to run in the background
+	// and collect more results for filtering while the checker is running.
+	rdc.taskrunner.Schedule(func(ctx context.Context) error {
+		return rdc.runDispatch(ctx, collected[len(collected)-1].AfterResponseCursor)
+	})
+
+	return nil
+}
+
+func (rdc *rdc) runChecker(ctx context.Context, collected []*v1.DispatchLookupResources2Response) error {
+	rdc.lock.Lock()
+	if rdc.ci.limits.hasExhaustedLimit() {
+		rdc.lock.Unlock()
+		return nil
+	}
+	rdc.lock.Unlock()
+
+	checkHints := make([]*v1.CheckHint, 0, len(collected))
+	resourceIDsToCheck := make([]string, 0, len(collected))
+	for _, resource := range collected {
+		resourceIDsToCheck = append(resourceIDsToCheck, resource.Resource.ResourceId)
+
+		checkHint, err := hints.HintForEntrypoint(
+			rdc.entrypoint,
+			resource.Resource.ResourceId,
+			rdc.parentRequest.TerminalSubject,
+			&v1.ResourceCheckResult{
+				Membership: v1.ResourceCheckResult_MEMBER,
+			})
```

Comment:
> Why does the hint need to be represented as N hint elements, instead of a more compact/normalized representation?

Because they are isolated hints

> Each one of the resources has the same entrypoint and the same terminal subject. 

Here, but that may not apply in the future. This is a generalized hint system.

> Could we normalize this information, and if necessary, have the client-side derive these N hints?

Yes, I could have each hint support taking in multiple resource IDs, but that would require more complicated code, so its a tradeoff

Reply:
>Because they are isolated hints

I don't understand what this means. Please elaborate.

> Yes, I could have each hint support taking in multiple resource IDs, but that would require more complicated code, so its a tradeoff

My tests indicate that under an LR2 workload, SpiceDB is spending 32% of CPU time in GC (21-23% with my optimizations in https://github.com/authzed/spicedb/pull/1989) . That's almost a fourth of CPU time collecting garbage (❗). And the vast majority of it is proto allocations.

That CPU time that goes to GC means less CPU time for the massive goroutine fan out LR2 tends to do with certain data shapes and sizes. In turn, it increases tail latencies.

My advice is to make the tradeoff and reduce allocations, the impact cannot be ignored and is globally an issue across all API endpoints. This is _very relevant_ for deployments that have low CPU requests. Feel free to do in a follow up PR so this one does not stay around opened any longer.

Reply:
Okay